{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def composition(k, text):\n",
    "    \"\"\"\n",
    "    Assuming k is an integer and text is a string of nucleotides, return the list of all k-mers in text (including repeated k-mers).\n",
    "    >>> composition(5, \"CAATCCAAC\")\n",
    "    ['CAATC', 'AATCC', 'ATCCA', 'TCCAA', 'CCAAC']\n",
    "    \"\"\"\n",
    "    kmers = []\n",
    "    length = len(text)\n",
    "    for i in range(0, length-k+1):\n",
    "        kmers.append(text[i:i+k])\n",
    "    #print(*kmers, sep=\" \")\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PathToGenome(path):\n",
    "    \"\"\"\n",
    "    Assuming the input is a sequence path of n k-mers split with spaces, the consecutive ones of which shares an overlap whose length is k-1;\n",
    "    return an assembled string.\n",
    "    >>> PathToGenome(\"ACCGA CCGAA CGAAG GAAGC AAGCT\")\n",
    "    ACCGAAGCT\n",
    "    \"\"\"\n",
    "    kmers = path.split() if type(path) == str else path\n",
    "    return kmers[0] + ''.join(kmer[-1] for kmer in kmers[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Overlap(patterns):\n",
    "    \"\"\"\n",
    "    Assuming the input is a string of k-mers split with spaces, return the overlap graph in the form of an adjacency list\n",
    "    >>> Overlap(\"ATGCG GCATG CATGC AGGCA GGCAT GGCAC\")\n",
    "    {'GCATG': ['CATGC'],\n",
    "    'CATGC': ['ATGCG'],\n",
    "    'AGGCA': ['GGCAT', 'GGCAC'],\n",
    "    'GGCAT': ['GCATG']}\n",
    "    \"\"\"\n",
    "    kmers = patterns.split()\n",
    "    prefix = {kmer: kmer[:-1] for kmer in kmers}\n",
    "    suffix = {kmer: kmer[1:] for kmer in kmers}\n",
    "\n",
    "    adjacency = {kmer: [] for kmer in kmers}\n",
    "    for kmer in kmers:\n",
    "        for key, value in prefix.items():\n",
    "            if suffix[kmer] == value:\n",
    "                adjacency[kmer].append(key)\n",
    "    \n",
    "    # Remove all the keys in adjacency if its value is empty list\n",
    "    adjacency = {key: value for key, value in adjacency.items() if value}\n",
    "    # print('\\n'.join(f'{kmer}: {\" \".join(adj)}' for kmer, adj in adjacency.items()))\n",
    "    return adjacency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DeBruijn_text(k, text):\n",
    "    \"\"\"\n",
    "    Assuming k is an integer greater than 0 and text is a string whose length is greater than or equal to k,\n",
    "    the output is an adjacency list where all the identically labelled nodes are glued.\n",
    "    >>> DeBruijn_text(4, \"AAGATTCTCTAAGA\")\n",
    "    {'AAG': ['AGA', 'AGA'],\n",
    "    'AGA': ['GAT'],\n",
    "    'GAT': ['ATT'],\n",
    "    'ATT': ['TTC'],\n",
    "    'TTC': ['TCT'],\n",
    "    'TCT': ['CTC', 'CTA'],\n",
    "    'CTC': ['TCT'],\n",
    "    'CTA': ['TAA'],\n",
    "    'TAA': ['AAG']}\n",
    "    \"\"\"\n",
    "    length = len(text)\n",
    "    kmers = [text[i:i+k-1] for i in range(0, length-k+2)]\n",
    "    adjacency = {kmer: [] for kmer in kmers}\n",
    "    for i in range(0, len(kmers)-1):\n",
    "        adjacency[kmers[i]].append(kmers[i+1])\n",
    "    \n",
    "    # Remove all the keys in adjacency if its value is empty list\n",
    "    adjacency = {key: value for key, value in adjacency.items() if value}\n",
    "\n",
    "    #print('\\n'.join(f'{kmer}: {\" \".join(adj)}' for kmer, adj in adjacency.items()))\n",
    "    return adjacency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DeBruijn_pattern(patterns):\n",
    "    \"\"\"\n",
    "    Assume the input is a collection of k-mer patterns split with spaces,\n",
    "    the output is an adjacency list of the de Bruijn graph.\n",
    "    >>> DeBruijn_pattern(\"GAGG CAGG GGGG GGGA CAGG AGGG GGAG\")\n",
    "    {'GAG': ['AGG'],\n",
    "    'CAG': ['AGG', 'AGG'],\n",
    "    'GGG': ['GGG', 'GGA'],\n",
    "    'AGG': ['GGG'],\n",
    "    'GGA': ['GAG']}\n",
    "    \"\"\"\n",
    "    patterns = patterns.split()\n",
    "    k = len(patterns[0])\n",
    "    adjacency = {kmer[:k-1]: [] for kmer in patterns}\n",
    "\n",
    "    for kmer in patterns:\n",
    "        adjacency[kmer[:k-1]].append(kmer[1:])\n",
    "    \n",
    "    #print('\\n'.join(f'{kmer}: {\" \".join(adj)}' for kmer, adj in adjacency.items()))\n",
    "    return adjacency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def EulerianCycle(graph):\n",
    "    \"\"\"\n",
    "    Assume the input is a strongly connected directed graph in the form of an adjacency list, the output is an Eulerian cycle.\n",
    "    >>> EulerianCycle(\n",
    "    '''\n",
    "    0: 3\n",
    "    1: 0\n",
    "    2: 1 6\n",
    "    3: 2\n",
    "    4: 2\n",
    "    5: 4\n",
    "    6: 5 8\n",
    "    7: 9\n",
    "    8: 7\n",
    "    9: 6\n",
    "    ''')\n",
    "    ['2', '6', '8', '7', '9', '6', '5', '4', '2', '1', '0', '3', '2']\n",
    "    \"\"\"\n",
    "    import random\n",
    "\n",
    "    # Re-format the input and generate a graph dictionary where the key is node and value is a list consisting of its adjacent nodes, e.g., graph = {\"a\": [\"b\", \"c\"]}\n",
    "    if type(graph) == str:\n",
    "        node_graph = [i for i in [item.strip() for item in graph.split(\"\\n\")] if i != \"\"]\n",
    "        graph = {str(key): list(map(str, value.split())) for key, value in (item.split(\": \") for item in node_graph)}\n",
    "\n",
    "    start = random.choice(list(graph.keys()))\n",
    "    cycle = [start]\n",
    "    while graph != {}:\n",
    "        if cycle[0] == cycle[-1] and cycle[-1] not in graph:\n",
    "            new_start = random.choice([node for node in cycle if node in graph.keys()])\n",
    "            cycle = cycle[cycle.index(new_start):] + cycle[1:cycle.index(new_start)+1]\n",
    "        else:\n",
    "            next_step = random.choice(graph[cycle[-1]])\n",
    "            graph[cycle[-1]].remove(next_step)\n",
    "            cycle.append(next_step)\n",
    "            graph = {key: value for key, value in graph.items() if value}\n",
    "    #print(\" \".join(str(x) for x in cycle))\n",
    "    return cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def EulerianPath(graph):\n",
    "    \"\"\"\n",
    "    Assume the input is a nearly balanced directed graph, the output is its Eulerian path.\n",
    "    >>> EulerianPath(\n",
    "    '''\n",
    "    0: 2\n",
    "    1: 3\n",
    "    2: 1\n",
    "    3: 0 4\n",
    "    6: 3 7\n",
    "    7: 8\n",
    "    8: 9\n",
    "    9: 6\n",
    "    ''')\n",
    "    ['6', '7', '8', '9', '6', '3', '0', '2', '1', '3', '4']\n",
    "    \"\"\"\n",
    "    import random\n",
    "    from collections import Counter\n",
    "\n",
    "    # Re-format the input and generate a graph dictionary where the key is node and value is a list consisting of its adjacent nodes, e.g., graph = {\"a\": [\"b\", \"c\"]}\n",
    "    if type(graph) == str:\n",
    "        node_graph = [i for i in [item.strip() for item in graph.split(\"\\n\")] if i != \"\"]\n",
    "        graph = {str(key): list(map(str, value.split())) for key, value in (item.split(\": \") for item in node_graph)}\n",
    "    nodes = list(set([node for node in graph.keys()] + sum(graph.values(), [])))\n",
    "\n",
    "    # Find the imbalanced nodes, and make them the start and the end of the Eulerian path\n",
    "    for node in nodes:\n",
    "        if node not in graph:\n",
    "            graph[node] = []\n",
    "        indegree = dict(Counter(sum(graph.values(), [])))[node] if node in dict(Counter(sum(graph.values(), []))) else 0\n",
    "        outdegree = len(graph[node]) if graph[node] != [] else 0\n",
    "        if indegree < outdegree:\n",
    "            start = node\n",
    "        elif indegree > outdegree:\n",
    "            end = node\n",
    "\n",
    "    # Append one directed edge from the end to the start, and generate an Eulerian cycle\n",
    "    graph[end].append(start)\n",
    "    cycle = [start]\n",
    "    while graph != {}:\n",
    "        if cycle[0] == cycle[-1] and cycle[-1] not in graph:\n",
    "            new_start = random.choice([node for node in cycle if node in graph.keys()])\n",
    "            cycle = cycle[cycle.index(new_start):] + cycle[1:cycle.index(new_start)+1]\n",
    "        else:\n",
    "            next_step = random.choice(graph[cycle[-1]])\n",
    "            graph[cycle[-1]].remove(next_step)\n",
    "            cycle.append(next_step)\n",
    "            graph = {key: value for key, value in graph.items() if value}\n",
    "    \n",
    "    # Reshuffle the Eulerian cycle, remove the additional edge, and set the start and end\n",
    "    path = cycle[cycle.index(end)+1:] + cycle[1:cycle.index(end)+1]\n",
    "    while path[0] != start or path[-1] != end:\n",
    "        path = path[path.index(end)+1:] + path[:path.index(end)+1]\n",
    "    print(\" \".join(str(x) for x in path))\n",
    "    return path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def StringReconstruction(k, patterns):\n",
    "    \"\"\"\n",
    "    Assume the input is an integer k followed by a list of k-mers patterns,\n",
    "    the output is a string with k-mer composition equal to patterns.\n",
    "    >>> StringReconstruction(4, \"CTTA ACCA TACC GGCT GCTT TTAC\")\n",
    "    GGCTTACCA\n",
    "    \"\"\"\n",
    "    dB = DeBruijn_pattern(patterns)\n",
    "    path = EulerianPath(dB)\n",
    "    Text = PathToGenome(path)\n",
    "    return Text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BinaryString(k):\n",
    "    \"\"\"\n",
    "    Assume the input is an integer k > 0, the output is a list colletion of all possible binary k-mers\n",
    "    >>> BinaryString(3)\n",
    "    ['000', '001', '010', '011', '100', '101', '110', '111']\n",
    "    \"\"\"\n",
    "    if k == 1:\n",
    "        return [\"0\", \"1\"]\n",
    "    else:\n",
    "        return [kmer + bit for kmer in BinaryString(k-1) for bit in [\"0\", \"1\"]]\n",
    "\n",
    "def CircularString(k):\n",
    "    \"\"\"\n",
    "    Assume the input is an integer k > 0, the output is a list of unique k-universal circular string\n",
    "    and the total number of all unique circular strings.\n",
    "    >>> CircularString(3)\n",
    "    (['10001110', '00010111'], 2)\n",
    "    \"\"\"\n",
    "    # Generate a binary k-mer list that includes all the possibilities\n",
    "    binary_kmer = BinaryString(k)\n",
    "    cycles = []\n",
    "    for i in range(1000):\n",
    "        path = EulerianCycle(DeBruijn_pattern(\" \".join(binary_kmer)))\n",
    "        Text = PathToGenome(path)\n",
    "        cycles.append(Text[:-k+1]) # To remove the overlap whose length is k-1\n",
    "    \n",
    "    # Make a list composed of unique k-universal circular strings\n",
    "    cycles = list(set(cycles)) # Make every element in the cycles unique\n",
    "    sorted_substrings = {}\n",
    "\n",
    "    for cycle in cycles:\n",
    "        double_cycle = cycle + cycle # Double the cycle to make a linearized string\n",
    "        substrings = [double_cycle[i:i+len(cycle)] for i in range(len(cycle))] # Create a list which includes all the k-mers whose length is the length of cycle\n",
    "        sorted_substrings[\" \".join(sorted(substrings))] = cycle # Discard all the strings that have the same k-mers\n",
    "\n",
    "    return list(sorted_substrings.values()), len(sorted_substrings)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
