{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prerequisites\n",
    "\n",
    "def hammingdistance(seq1, seq2):\n",
    "    seq1 = list(seq1)\n",
    "    seq2 = list(seq2)\n",
    "    hmgdist = 0\n",
    "    for i in range(len(seq1)):\n",
    "        if seq1[i] != seq2[i]:\n",
    "            hmgdist += 1\n",
    "    return hmgdist\n",
    "\n",
    "def Suffix(Pattern):\n",
    "    Pattern = list(Pattern)\n",
    "    Pattern.pop(0)\n",
    "    return \"\".join(Pattern)\n",
    "\n",
    "def Neighbors(Pattern, d):\n",
    "    Nucleotides = [\"A\", \"C\", \"T\", \"G\"]\n",
    "    if d == 0:\n",
    "        return Pattern\n",
    "    if len(Pattern) == 1:\n",
    "        return Nucleotides\n",
    "    \n",
    "    Neighborhood = []\n",
    "    SuffixNeighbors = Neighbors(Suffix(Pattern), d)\n",
    "    for i in SuffixNeighbors:\n",
    "        if hammingdistance(Suffix(Pattern), i) < d:\n",
    "            for nt in Nucleotides:\n",
    "                Neighborhood.append(nt + i)\n",
    "        else:\n",
    "            PatternList = list(Pattern)\n",
    "            Neighborhood.append(PatternList[0] + i)\n",
    "    \n",
    "    return Neighborhood\n",
    "\n",
    "def ReverseComplement(Pattern):\n",
    "    Pattern = Pattern[::-1]\n",
    "    Pattern_list = list(Pattern)\n",
    "    for i in range(0, len(Pattern)):\n",
    "        if Pattern_list[i] == \"A\":\n",
    "            Pattern_list[i] = \"T\"\n",
    "        elif Pattern_list[i] == \"T\":\n",
    "            Pattern_list[i] = \"A\"\n",
    "        elif Pattern_list[i] == \"C\":\n",
    "            Pattern_list[i] = \"G\"\n",
    "        elif Pattern_list[i] == \"G\":\n",
    "            Pattern_list[i] = \"C\"\n",
    "    result = \"\"\n",
    "    for i in Pattern_list:\n",
    "        result += i\n",
    "    return result\n",
    "\n",
    "def minHamDist(pattern1, pattern2):\n",
    "    # Pattern1 should be shorter than pattern2\n",
    "    if len(pattern1) > len(pattern2):\n",
    "        pattern1, pattern2 = pattern2, pattern1\n",
    "    len1 = len(pattern1)\n",
    "    len2 = len(pattern2)\n",
    "    minDist = len(pattern1)\n",
    "\n",
    "    for i in range(0, len2-len1+1):\n",
    "        kmer = pattern2[i:i+len1]\n",
    "        if hammingdistance(pattern1, kmer) < minDist:\n",
    "            minDist = hammingdistance(pattern1, kmer)\n",
    "    \n",
    "    return minDist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# (k, d)-motif: a k-mer if it appears in every string from Dna with at most d mismatches\n",
    "# Brute force, generating all possible k-mers\n",
    "\n",
    "def MotifEnumeration(Dna, k, d):\n",
    "    Patterns = []\n",
    "    Dna = Dna.split()\n",
    "\n",
    "    line1 = Dna[0]\n",
    "    kmerlist_line1 = []\n",
    "    for i in range(0, len(line1)-k+1):\n",
    "        kmer = line1[i:i+k]\n",
    "        kmerlist_line1.append(kmer)\n",
    "\n",
    "   # Creating a list of all k-mer neighbor differing from Pattern by at most d mismatches \n",
    "    all_neighbors = []\n",
    "    all_neighbors += kmerlist_line1\n",
    "    for kmer in kmerlist_line1:\n",
    "        neighbors = Neighbors(kmer, d)\n",
    "        all_neighbors += neighbors\n",
    "    all_neighbors = list(set(all_neighbors))\n",
    "\n",
    "    # To make sure if there is one hit with maximum hamming distance of d on every single line of genomic DNA\n",
    "    for neighbor in all_neighbors:\n",
    "        criteria = 0\n",
    "        for line in Dna[1:]:\n",
    "            minDist = minHamDist(neighbor, line)\n",
    "            if minDist <= d:\n",
    "                criteria += 1\n",
    "        if criteria == len(Dna[1:]):\n",
    "            Patterns.append(neighbor)\n",
    "\n",
    "    return list(set(Patterns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
